### PR TITLE
Add feature to list all templates

### PR DESCRIPTION
**Pull request by abrahamks**
I added a new class `Template` and a temporary method `request_get` for module `Request` to avoid a breaking change. Ideally, this method should be abstracted into a `request` method with an extra parameter which accept HTTP Verb

### CODE REVIEW

**Comment by rajumsys** [comment]
Awesome @abrahamks. Thanks for the PR. I'll have a look at it soon. 

Thanks

File: lib/sparkpost/request.rb [hunk 1]
```diff
@@ -17,7 +17,20 @@ def request(url, api_key, data)
       }
       req = Net::HTTP::Post.new(uri.path, headers)
       req.body = data.to_json
+      process_response(http.request(req))
+    end
 
+    def request_get(url, api_key, _data)
```
Comment (rajumsys) [inline lib/sparkpost/request.rb]: `request` and `request_get` now doing lot of redundant tasks. Can we update it to re-use codes as much as possible?

File: lib/sparkpost/template.rb [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+    def initialize(api_key, api_host)
+      @api_key = api_key
+      @api_host = api_host
+    end
+
+    def endpoint
+      @endpoint ||= @api_host.concat('/api/v1/templates')
+    end
+
+    def list_all_templates
```
Comment (rajumsys) [inline lib/sparkpost/template.rb]: Let's name it just `list` as our python library does.

File: lib/sparkpost/request.rb [hunk 3]
```diff
@@ -6,21 +6,39 @@
 
 module SparkPost
   module Request
-    def request(url, api_key, data)
+    def request(verb, url, api_key, data)
```
Comment (rajumsys) [inline lib/sparkpost/request.rb]: Thanks for the refactoring. The only concern is that we've suggested, at least a user, to use `request` method before we introduced `send_payload`. I think if we make `verb` as an optional argument (make it last arg) and default to POST, it'll be backward compatible. What do you think?

Comment (rajumsys) [inline lib/sparkpost/request.rb]: Well, I'm not sure what `POST

Comment (abrahamks) [inline lib/sparkpost/request.rb]: @rajumsys Thanks for the feedback. Sure, I'll refactor `request` method to be backward compatible.

File: examples/template/list_all_templates.rb [hunk 4]
```diff
@@ -0,0 +1,8 @@
+#!/usr/bin/env ruby
+
+require_relative '../../lib/sparkpost'
+
+sp = SparkPost::Client.new # api key was set in ENV
+templates = sp.template.list_all_templates
```
Comment (rajumsys) [inline examples/template/list_all_templates.rb]: `sp.template.list` now.

Comment (abrahamks) [inline examples/template/list_all_templates.rb]: Yes, I will change this one too

**Comment by rajumsys** [comment]
@abrahamks  thanks for the update. All looks good. Seems it now has conflicts with master. If you rebase with master, I think we're good to merge it. Thanks for great PR :+1:

**Pull request merged**